pow/src/main/java/tech/pegasys/teku/pow/Web3jEth1Provider.java

```diff
@@ -162,4 +190,46 @@ public Web3jEth1Provider(final Web3j web3j, final AsyncRunner asyncRunner) {
               return (List<EthLog.LogResult<?>>) (List) logs;
             });
   }
+
+  @Override
+  public SafeFuture<Boolean> validate() {
+    if (validating.compareAndSet(false, true)) {
+      LOG.info("Validating endpoint {} ...", this.id);
+      return getChainId()
+          .thenAccept(
+              chainId -> {
+                if (chainId.intValueExact() != Constants.DEPOSIT_CHAIN_ID) {
+                  STATUS_LOG.eth1DepositChainIdMismatch(
+                      Constants.DEPOSIT_CHAIN_ID, chainId.intValueExact(), this.id);
+                  throw new RuntimeException("Wrong Chainid");
+                }
+              })
+          .thenCompose(__ -> this.ethSyncing())
+          .handleComposed(
+              (syncing, err) -> {
+                SafeFuture<Boolean> futureReturn = new SafeFuture<>();
+                if (err != null) {
+                  LOG.warn(
+                      "Endpoint {} is INVALID | {}",
+                      this.id,
+                      Throwables.getRootCause(err).getMessage());
+                  updateLastValidation(Result.failed);
+                  futureReturn.complete(Boolean.FALSE);
+                } else if (syncing) {
+                  LOG.warn("Endpoint {} is INVALID | Still syncing", this.id);
+                  updateLastValidation(Result.failed);
+                  futureReturn.complete(Boolean.FALSE);
+                } else {
+                  LOG.info("Endpoint {} is VALID", this.id);
+                  updateLastValidation(Result.success);
+                  futureReturn.complete(Boolean.TRUE);
+                }
+                validating.set(false);
+                return futureReturn;
+              });
```

Comment:
For safety, we should set validating to false in an `.alwaysRun` block.  That way we guarantee validating gets set back to false even if an unexpected exception gets thrown in this handling code.
```suggestion
                return futureReturn;
              })
              .alwaysRun(() -> validating.set(false));
```

`.alwaysRun` here is like a `finally` block in a try/catch.